Binary_Search_1_solns.py

```diff
@@ -0,0 +1,88 @@
+# Time Complexity : O(log(m*n))
+# Space Complexity : O(1)
+# Did this code successfully run on Leetcode : Yes
+# Any problem you faced while coding this : Yes I was dividing row by m not n
+
+class Solution:
+    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
+        m = len(matrix)
+        n = len(matrix[0])
+        low = 0
+        high = m*n-1
+        
+        while low <= high:
+            mid = (low + high) // 2
+            row = mid // n
+            col = mid % n
+            
+            if matrix[row][col] == target:
+                return True
+            elif target < matrix[row][col]:
+                high = mid - 1
+                
+            else:
+                low = mid + 1                
+        
+        return False
+
+
+# Time Complexity : O(log(n))
+# Space Complexity : O(1)
+# Did this code successfully run on Leetcode : Yes
+# Any problem you faced while coding this : No
+
+class Solution:
+    def search(self, nums: List[int], target: int) -> int:
+        low = 0
+        high = len(nums) - 1
+        
+        while low <= high:
+            mid = low + (high - low) // 2
+            
+            if nums[mid] == target:
+                return mid
+            
+            if nums[low] <= nums[mid]:
+                if target < nums[mid] and target >= nums[low]:
+                    high = mid - 1
+                    
+                else:
+                    low = mid + 1
+                    
+            elif nums[mid] <= nums[high]:
+                if target > nums[mid] and target <= nums[high]:
+                    low = mid + 1
+                else:
+                    high = mid - 1
+                    
+        return -1
+
+# Time Complexity : O(logm + logn) ||  where m is the number of steps required for the target element to fall within range of the subarray. n is number of elements within the range
+# Space Complexity : O(1)
+# Did this code successfully run on Leetcode : Yes
+# Any problem you faced while coding this : Nope ... A bit confused about the time complexity
```

Comment:
Correct. Since `n` will be very large than `m`, we can say the time complexity as `O(logn)`  where n is the range between low and high.